Fixed ResourceSaver::save not working with GDScript EditorPlugin

gtsiam opened the pull request:
Since this is my first contribution in a public repository, please note that I don't completely understand how git works, so expect mistakes in that area. With that out of the way:

**Initial behavior:** Calls to `ResourceSaver::save(path, resource)` with an edited object from a GDScript editor plugin resulted in the resource being null when it arrived at the C++ function from GDScript.

**Issue:** The edited object was passed to editor plugins through the edit() function as a Variant, which was constructed by the Variant::Variant(Object*) constructor. However the constructor did not handle the case of the passed Object being a Reference, meaning that the variant was not considered to hold a Reference at that point (Nor a Resource, in the case of `ResourceSaver::save`) so I added a special case for that.

Also, I removed the (already commented out), `operator RefPtr()` in reference.h, which seems to have now been replaced by the more direct `operator Variant()`

**Fixes Issue:**#6603

gtsiam commented:
Why the confused emoji? Could you please be more specific? In short, it just fixes the bug regarding ResourceSaver::save not working with EditorPlugins.

bojidar-bg commented:
@gtsiam I'm just not sure if that is to be fixed in Variant, or if edit() should be changed to take Variant instead of Object*. This is something the other core devs should decide though, I'm not sure what the original intention was when this code was written.

gtsiam commented:
It is kind of weird to explain the issue took me a few days to understand it myself, but here I go again:

The edit function (in GDScript editor plugins), accepts a variant, that is just how GDScript works. Upon debugging, I found that the variant passed to _ResourceSaver::save (in core/bind/core_bind.h:82) did not contain a reference, just an object. What I mean by that:

Variants, internally, have a `_data` union, which stores the variant's contents. For Objects and References, you can get (with `_get_obj_data()`) an `ObjData` struct, which contains two fields: `Object *obj`, and `RefPtr ref`. Issue is, that if you passed in the Variant constructor, an Object, even if it is of type Reference, the `ref` part of ObjData, would still be null, which means that when casting to the Resource `_ResourceSaver::save` expects, The `VariantCaster` (in core/method_bind.h), would think that they are incompatible, and return null.

gtsiam commented:
The engine crashed with target=release_debug, because I had forgotten to check if the p_object argument was null (which was fine in target=debug, for some reason?). Anyway, I added a check (and kind of messed up the commit author, but that should also be fixed now).

vnen commented:
I don't think that's the proper way to fix it. `ResourceSaver::save()` **does** work from GDScript, so the title may be a bit misleading. The problem I had is just that the object got from `EditorPlugin::edit()` does not work with `ResourceSaver::save()`. I believe it should be fixed somewhere in the EditorPlugin side, not changing how the Variant constructor works.

The only problem that I see is that `edit()` receives an Object, (which can be a Resource) but `ResourceSaver::save()` expects a `Ref<Resource>`. So maybe the plugin can receive a `Ref<Object>`, but I don't really know the implication of that. Or as @bojidar-bg said, the plugin should receive a Variant (this makes sense for me, considering it goes to a script).

gtsiam commented:
@vnen You do have a point about the title, I changed it.

However, I urge you to reconsider about not changing the constructor. Why not have it just be smart enough to realize when the passed object is a reference? It is possible to change just the `edit()` function, but that would be liable to cause code duplication, when someone else encounters this issue in the future, because internally, it is not just about `edit()` function, but about how Variants store `Ref<...>` and `Object` differently. And, to me at least, it does seem like a very hacky way of dealing with the problem..

vnen commented:
I don't know, I find this patch weird. It basically takes all the Reference types (which includes all Nodes and Resources) and wrap it into a `Ref<>` object. I can't tell the implications of this in the general use or how it'll interfere with reference counting. You even removed the constness of the constructor argument.

I would like that core devs took a look into this to see if it's correct.

gtsiam committed:
Variant(Object*) only creates REF when the object is referenced at least once

gtsiam commented:
You were right, there was an issue with reference counting, however I have spent the past few days trying to figure out a (proper) way around that, and I am sorry, but I give up... The reference situation inside godot is a bit of a mess. One thing that you really need to fix is Reference objects that are never being placed inside of a `Ref<>` (Memory leaks maybe?)

I will, though upload my code where I left off. It does fix the ResourceSaver::save problem, but it is a hacky, in my opinion, way of fixing it - But such a hacky way of fixing the same problem already exists ([here](https://github.com/godotengine/godot/blob/a5d500f0233913fe7622434225c5dc73ebaccd1e/modules/gdscript/gd_function.cpp#L255)), so you might want to accept it, either way.

As for how this change would affect other parts of the engine, it shouldn't affect anything. The only thing it could potentially do, is to make some variants that previously didn't count towards the reference counter, count, thus prolonging their life time - not really a bad thing.

As for the constructor constness, it never really existed - the Object was going through a const_cast. I don't know about you, but I would count that kind of constness as a bug, since variant objects can and will almost certainly be modified at some point.

reduz commented:
My take on solving this problem is making:
Ref::Ref(const Variant &p_variant)
explicit, this way we can avoid conversions.. no idea what effect it will have, but we could give it a try..

gtsiam commented:
I'll be honest, I am not really sure how changing the `Ref::Ref (const Variant&)` constructor from the default implicit to explicit would help the problem. Could you please clarify?

akien-mga commented:
Quoting @reduz from IRC:
```
<reduz> Akien: the fix is not good, it hampers performance a lot.  I just couldnt figure out a good way to fix it yet
<reduz> tried to make the constructor in Reference explicit but then it doesnt work and i dont quite understand why (or probably I do, just have no idea how to fix)
<reduz> Akien: still, this bug results from a very very corner use case that could be fixed directly in the editor instead
```

So closing for now, issue #6603 is still there for further discussion about what could be a proper fix.

gtsiam commented:
@akien-mga I know this is a very late response, but here I go:

I did come up with a (possibly) better fix a while ago, but did not have time to implement it earlier (check [gtsiam/godot](https://github.com/gtsiam/godot)):

Added two methods to the `Variant` class:

- `RefPtr Variant::try_get_ref(bool init_ref = false)`:
Tries to convert the `Reference` object stored in the `Variant`, to a `RefPtr`, and returns it, or an empty `RefPtr` if it fails. It, however, does not assign it to the Variant, so us not to prolong the `Reference`'s life longer than needed. The `bool init_ref` parameter controls whether or not to create the reference, if no other exists. Setting it to true means that, if no other REF is created to the referenced object, then the the object will be tied to the returned reference's lifetime (Usually undisired behaviour)

- `RefPtr Variant::try_make_ref(bool init_ref = false)`
Actually assigns the RefPtr to the variant, so it will from now on act as a normal REF<Reference> Variant. In case of success, is_ref() will return true in the future.`init_ref` works same as above.

Also added specialized `VariantCaster<T>` structs, where `T` is for `Ref<T>` objects, where the cast method calls the `try_get_ref()` function. Functions that accept RefPtr as a parameter will not work the same way in this corner case, but seriously, why would you call such a function from GDScript? (If there is such a function, please tell me though...)
